Automatic Xcode management in fastlane.ci

KrauseFx opened the pull request:
### Context

`fastlane.ci` should automatically manage Xcode installations for the user 🎉 

- Implements https://github.com/fastlane/ci/issues/446
- Implements [update_xcode_installation.md](https://github.com/fastlane/ci/blob/master/docs/use_cases/update_xcode_installation.md)
- Improve `xcode-install` API to use with fastlane.ci with https://github.com/KrauseFx/xcode-install/pull/275
- Introduced `.xcode-version` file with https://github.com/Themoji/ios/pull/32
- Add support for `.xcode-version` file to _fastlane_ with https://github.com/fastlane/fastlane/pull/12479
- New download system for Xcode installations https://github.com/KrauseFx/xcode-install/issues/276 to follow along progress (implemented with https://github.com/KrauseFx/xcode-install/pull/278)

### Changes introduced with this PR:

- Interact with [xcode-install](https://github.com/krausefx/xcode-install)
- List installed and available Xcode versions
- Store used Xcode version as part of build details
- Introduces the `.xcode-version` file
- Set Xcode path based on version for each build automatically if `.xcode-version` file exists
- Add detection if `.xcode-version` file Xcode version is already installed, if it's available to install, or if it's invalid
- Follow-along progress of Xcode download
- Trigger Xcode installations from a Web UI
- Use of user's Apple ID and verification it's valid
- Report `Installing Xcode` status to GitHub API

### Open tasks

- [x] Build a fastlane action that verifies the versions match (with https://github.com/fastlane/fastlane/pull/12479)
- [x] Decide on how we can identify an Xcode version in a unique way (version number, build number)
- [x] Store used Xcode version on every build, default to current default
- [x] Implement installing of new Xcode versions
- [x] Decide on a flow on how new Xcode versions are installed on the fly, probably by pausing a build once the `.xcode-version` file was loaded
- [x] Manage iTunes Connect credentials for managing Xcode installations (discussion with https://github.com/fastlane/ci/issues/851, implementation with https://github.com/fastlane/ci/pull/854)
- [x] Show in-progress installations somewhere
- [x] Block current build and put it back in the queue during Xcode installation
- [x] Attach `success_block` to already running Xcode downloads if multiple builds wait for it

### Separate PRs/feature:

- Implement asking for the root password as part of the installation (via https://github.com/KrauseFx/xcode-install/issues/279) -> https://github.com/fastlane/ci/issues/902
- Allow setting of a system default Xcode version -> https://github.com/fastlane/ci/issues/870
- Once a new Xcode version was released, install it automatically, create a new branch for every PR-trigger-based project and test if the tests are still green -> https://github.com/fastlane/ci/issues/903
- Decide if and how we're gonna use `@_installer.rm_list_cache` -> https://github.com/fastlane/ci/issues/904

---

### List installed and available Xcode versions & download in progress

<img width="1024" alt="screenshot 2018-05-11 09 30 38" src="https://user-images.githubusercontent.com/869950/39912135-0126af5a-54fe-11e8-8e17-0c86d1f2128f.png">

### Store used Xcode version as part of build details
<img width="738" alt="screenshot 2018-05-07 03 47 49" src="https://user-images.githubusercontent.com/869950/39705486-daea5644-520e-11e8-9134-0b7f5e66d142.png">

### Show `Install Xcode as GitHub build status`
<img width="702" alt="screenshot 2018-05-09 06 42 19" src="https://user-images.githubusercontent.com/869950/39827629-d1705b76-53b8-11e8-9850-ee9f70b99232.png">

KrauseFx committed:
Started with Xcode management doc

KrauseFx committed:
Initial Xcode integration to list installed/available Xcode versions

KrauseFx committed:
Add dummy action buttons

minuscorp commented on app/services/xcode_manager_service.rb:
```diff
@@ -0,0 +1,42 @@
+  class XcodeManagerService
+    def installer
+      if @_installer.nil?
+        @_installer = XcodeInstall::Installer.new
```
`@_installer ||= XcodeInstall::Installer.new`?

KrauseFx replied:
We're already checking for `nil?` so this won't be necessary 👍

minuscorp replied:
I mean, that `nil?` check is unnecessary if you just `return @_installer ||= XcodeInstall::Installer.new`?

KrauseFx committed:
Add detecting and storing of Xcode version for builds using `.xcode-version` file & setting of DEVELOPER_DIR path

KrauseFx committed:
Bugfix

KrauseFx committed:
Add `xcode-install` dependency to Gemfile

KrauseFx commented:
Unfortunately https://developer.apple.com/download/more/ doesn't expose the exact build number of each Xcode release, so we can't use that to identify and download an Xcode release, will try to come up with something better

KrauseFx committed:
Add docs for `.xcode-version` file

KrauseFx committed:
Specify file location

KrauseFx committed:
Add link to xcode-version.md

KrauseFx committed:
Add information on how to parse & compare `.xcode-version` files

KrauseFx committed:
Add more code comments

KrauseFx committed:
Add information about trailing new line

KrauseFx committed:
Add initial Xcode service specs

KrauseFx committed:
Extend XcodeManagerService & add tests

KrauseFx committed:
Always stored the used Xcode version on every build & refactor

KrauseFx committed:
Initial implementation for installing Xcode from fastlane.ci

KrauseFx committed:
Use built-in methods to check for available Xcode versions

KrauseFx committed:
Keep record of what versions of Xcode we're installing & show them

KrauseFx committed:
Add more tests for Xcode service

KrauseFx commented:
Yeah, that's a great idea. It does seem however that it's not possible to download older beta releases, as Apple removes them from their server: https://github.com/fastlane/ci/pull/819/files#diff-0a8ab3aef33f23614a28a3c405f10991R43

KrauseFx committed:
Update GitHubService to support custom descriptions & add :installing_xcode status

KrauseFx committed:
Implement on-the-fly installation of Xcode & automatic re-queueing of build after success

taquitos commented on app/features/build_runner/build_runner.rb:
```diff
@@ -173,18 +173,26 @@ def pre_run_action(&completion_block)
         if checkout_success
-          setup_tooling_environment
-          setup_build_specific_environment_variables
+          if setup_tooling_environment # see comment for `#setup_tooling_environment` method
```
what happens if this return false? What kind of recovery do we need/error messaging?

taquitos commented on app/services/code_hosting/git_hub_service.rb:
```diff
@@ -187,7 +187,15 @@ def set_build_status!(repo: nil, sha: nil, state: nil, target_url: nil, descript
+        "success",
+        "ci_problem",
+        "missing_fastfile",
+        "installing_xcode"
```
❤️

KrauseFx committed:
Show download progress in Xcode dashboard

AndrewMcBurney commented on app/features/build_runner/build_runner.rb:
```diff
@@ -170,17 +170,33 @@ def checkout_sha(&completion_block)
       checkout_sha do |checkout_success|
         if checkout_success
-          setup_build_specific_environment_variables
+          if setup_tooling_environment # see comment for `#setup_tooling_environment` method
```
Could we join the two conditions `checkout_success` and `setup_tooling_environment` please?

i.e.,

```ruby
if checkout_success && setup_tooling_environment
  setup_build_specific_environment_variables
  completion_block.call(checkout_success)
else
  # ...
end
```

KrauseFx replied:
Yep, will do, thanks

AndrewMcBurney commented on app/features/build_runner/build_runner.rb:
```diff
@@ -170,17 +170,33 @@ def checkout_sha(&completion_block)
+    #         It's critical that the `setup_tooling_environment` method
+    #         added the same build runner onto the work queue again
+    #         Check out the `fastlane_build_runner` implementation for more details
+    def setup_tooling_environment
```
This is a nitpick, but do you think we could suffix this method with a `?` to make it apparent it returns a `Boolean` where it's being used?

KrauseFx replied:
Yes, great idea 👍

AndrewMcBurney commented on app/features/xcode_manager/views/xcode_manager.erb:
```diff
@@ -0,0 +1,108 @@
+                <%= xcode_version.version %>
+              </td>
+              <td>
+                <%= xcode_version.installed? ? "✅" : "👎️" %>
```
💯

AndrewMcBurney commented on app/services/code_hosting/git_hub_service.rb:
```diff
@@ -187,26 +187,43 @@ def set_build_status!(repo: nil, sha: nil, state: nil, target_url: nil, descript
       repo = repo.split("/")[-2..-1].join("/")
 
       if description.nil?
         description = "All green" if state == "success"
```
Do you think we could do a `case` statement here instead of having all the `if state == ""` appending the assignment to `description`?

i.e.,

```ruby
description =
  case state
  when "success"
    "All green"
  when "pending"
    "Still running"
  when "installing_xcode"
    "Installing Xcode"
  # ...
  end
```

KrauseFx replied:
Yeah, will do

AndrewMcBurney commented on app/services/code_hosting/git_hub_service.rb:
```diff
@@ -187,26 +187,43 @@ def set_build_status!(repo: nil, sha: nil, state: nil, target_url: nil, descript
+      if state == "missing_fastfile" || state == "ci_problem"
+        state = "failure"
+      end
+      if state == "installing_xcode"
```
Could we merge these two `if` statements and do an `elsif` for the second condition, please?

AndrewMcBurney commented on app/services/xcode_manager_service.rb:
```diff
@@ -0,0 +1,170 @@
+    #      "https://developer.apple.com/devcenter/download.action?path=/Developer_Tools/Xcode_6.4/Xcode_6.4.dmg",
+    #     @version=Gem::Version.new("6.4")>,
+    #
+    # the resulting list is sorted with the most recent release as first element
```
Nice documentation here! 👍

AndrewMcBurney commented on app/services/xcode_manager_service.rb:
```diff
@@ -0,0 +1,170 @@
+        raise "No `XCODE_INSTALL_USER` ENV variable provided, please provide one when launching up the fastlane.ci"
+      end
+
+      self.installations_in_progress = {}
```
This is a nitpick, but could we change this to be `@installations_in_progress = {}` to be consistent with how we're setting variables with `attr_accessor`s in other clases?

KrauseFx replied:
Yep

AndrewMcBurney commented on app/services/xcode_manager_service.rb:
```diff
@@ -0,0 +1,170 @@
+    end
+
+    # A shared reference to the `XcodeInstall::Installer` object we use
+    def installer
```
If we're deleting the comment `# @_installer.rm_list_cache`, could we please use a memoized method?

```ruby
def installer
  @_installer ||= Installer.new
end
```

KrauseFx replied:
Yeah, it's an open task to take a look at if we should call this method

taquitos commented on docs/XcodeManagement.md:
```diff
@@ -0,0 +1,26 @@
+
+## Opportunity
+
+- Managing Xcode installations is one of the top pain points for all generic CI systems. Mobile-optimzied CI systems can significantly help the user by automating the management of Xcode installations
```
💯

KrauseFx committed:
Rename setup_tooling_environment method

KrauseFx committed:
Make use of `case when` syntax in GitHub service

KrauseFx committed:
Use @ syntax instead of `self`

KrauseFx committed:
Merge branch 'master' into xcode

# Conflicts:
#	app/features/global/navigation.erb
#	app/services/services.rb
#	launch.rb

KrauseFx committed:
Rename Apple ID controller

KrauseFx committed:
Merge branch 'master' into xcode

KrauseFx committed:
Make use of Apple ID for Xcode installation

- Add placehodlers (HODL) for error handling when Apple ID can't be found
- Add tests

KrauseFx committed:
Add more tests for Xcode installation

KrauseFx committed:
Show Xcode installation trigger as part of build output

As part of this small refactor in the output row class

KrauseFx committed:
Address feedback from @AndrewMcBurney

KrauseFx committed:
Simplify build_runner `if` statement

KrauseFx committed:
Address feedback from @snatchev

KrauseFx committed:
Revert "Simplify build_runner `if` statement"

This reverts commit 006c0c8b1aa0ad9a8c3d66b7e5eebcca7648571a.

KrauseFx committed:
Fix code style

KrauseFx committed:
Update xcode-install dependency to support fastlane.ci

snatchev reviewed (approved):
This is a big PR and I want to get it merged and unblock more development :)

@KrauseFx lets sync about a follow up PR. I would like to extract some of the functionality in `setup_tooling_environment?` so that it can be re-used in other places, and also reduce the scope of that query method.
